File: leetcode/257-binary-tree-paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
  if root.nil?
    []
  else
    left = binary_tree_paths root.left
    right = binary_tree_paths root.right
    (left.empty? and right.empty?) ? [root.val.to_s] : (left + right).map {|s| "#{root.val}->#{s}"}
  end
end

File: leetcode/257-binary-tree-paths.cpp
/*
Given a binary tree, return all root-to-leaf paths.

For example, given the following binary tree:

   1
 /   \
2     3
 \
  5
All root-to-leaf paths are:

["1->2->5", "1->3"]
Credits:
Special thanks to @jianchao.li.fighter for adding this problem and creating all test cases.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        string s;
        vector<string> results;
        collect(root, s, results);
        return results;
    }

private:

    void collect(TreeNode *root, string s, vector<string> &results) {
        if (root == nullptr) {
            return;
        }
        stringstream ss;
        ss << root->val;
        s.append(ss.str());
        if (root->left != nullptr || root->right != nullptr) {
            s.append("->");
            collect(root->left, s, results);
            collect(root->right, s, results);
        } else {
            results.push_back(s);
        }
    }
};
